### PR TITLE
Client tests

### PR DESCRIPTION
**Pull request by uanabitarte**
Here is the first try of the client's tests. Some issues that needs to be solved:
- I think getting a timestamp to create client names ('client'+timestamp)  might cause unique constraint problems.
- I am having trouble creating Script fixtures, so the tests to check Invalid scripts are commented.
- The test to check if a job is enqueued cannot be deleted is missing, as I do not now how to enqueue a job from the test class.

Other than that, I would appreciate if you review the PR and suggest any improvement or corrections.

### CODE REVIEW

File: config/packages/test/doctrine.yaml
Comment (elacunza) [inline config/packages/test/doctrine.yaml:9]: Missing newline for last line

File: phpunit.xml.dist
Comment (elacunza) [inline phpunit.xml.dist:40]: Do we need this? Otherwise remove?

File: src/Api/Test/ClientTest.php
Comment (elacunza) [inline src/Api/Test/ClientTest.php:110]: Excess of blank lines

File: src/Service/JobService.php
Comment (elacunza) [inline src/Service/JobService.php:41]: Can we PR this independently, as it is not related to tests?

File: tests.sh
Comment (elacunza) [inline tests.sh:1]: I suggest calling this run-tests.sh that should be more self-explanatory?

**Review by elacunza — CHANGES_REQUESTED** [review]
Only picky comments from my part (sorry). Let's wait for the review by @igorbga

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:307]: Why do you use $timestamp as client id?

If you want to referene a non existing client just use a fixed high number or something like that. I know the timestamp is in fact a hi number but semantic also counts. Anyone seeing this code will think that someone the timestamp has something to do with what you are doing and is not. Or am I wrong?

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:309]: In other tests you are also checking the JSON's hydra error properties. Why not here?
Let's be consistent and to it the same way everywhere.

Maybe it's worth creating an "assertHydraError" function that receives the expected error code and message and groups your checks.

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:312]: Item?

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:324]: Well known client from a fixture instead?

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:344]: This test might be more interesting if you get 2 clients, one with the minimun (like the one you already have) and one with all parameters set.
This way it might also catch serialization related issues.

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:374]: Why not just append something to the original timestamped name? Otherwise you might collide with an already existing client name.

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:379]: You are not sending sshArgs, rsyncShortArgs etc...  and this is a PUT not a PATCH. Is this allowed?

It will delete/empty non provided properties or will it just leave them as they are? 

If it is the first one well that may make sense in the context of a REST PUT call but is a bit unexpected I'd rather we changed the API so that for a PUT all fields are mandatory. That change is out of the scope of this pull request but please file a bug/enhacement request so we don't lost track of this. And change this implementation so that when we make that change the test still works properly.

If it is the second one then this is not a proper REST PUT, it would be a PATCH. Please make sure that PUTs behave like PUTs should and not like PATCHes.

Apart from that I think that you should exercise changes in all fields of the client, not just some of them.

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:403]: Check for all changed fields

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:455]: Use a well known client from a fixture?

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:461]: Use timestamped name to avoid name collisions

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:490]: Hi number instead ?
We are using the UNEXISTING_ID many times so maybe we should take it to a constant.

File: src/Api/Test/ClientTest.php
Comment (igorbga) [inline src/Api/Test/ClientTest.php:515]: How many times have we already created clients? Maybe it's time to at least extract it to a "createClient" methos.

**Review by igorbga — CHANGES_REQUESTED** [review]
Not very drastic changes, but quite a lot of them.

First a very important stupid mantra that we should keep in mind:
**We don't program the tests so that they pass when everything is ok. We program the tests so that they don't pass when something is wrong**


In the function that you'll end up creating to generate client names, apart from the timestamp add a little random part to it and that way it would be highly unlikely to have a name collision.

From the point of view of the [suggested best practices for the Pyramid of Tests](https://martinfowler.com/articles/practical-test-pyramid.html) we are probably doing too many "End to end" tests but well, those are the only tests we have so better too much than too few.

File: src/Api/Test/BaseApiTestCase.php
Comment (igorbga) [inline src/Api/Test/BaseApiTestCase.php:40]:
```suggestion
        $clientName = 'client_'.$time->getTimestamp().'_'.rand(1000, 9999);
```

**Review by igorbga — CHANGES_REQUESTED** [review]
This is really an improvement in comparison with previous push but I think you should still give it another try.

In your createClient function you are dealing with 2 responsasbilities, you are creating the payload to send (the array serializable to JSON), and you are also doing the send itself.

I would separate both concerns.

You also have a quite messy signature for the method due to the fact that sometimes you just need a couple of parameters + "default" ones and sometimes you want to be more specific about a client.

In testing this is usually managed using [ObjectMother](https://martinfowler.com/bliki/ObjectMother.html) classes. Don't fear, this is just a fancy name for a factory class that is able to create objects.

This allows to improve the readibility of your tests quite a lot. For example the most basic test could look like:

```
$clientJson = ClientJsonMother::random();
$httpCient->request('POST, $url, $clientJson); 
$this->assertStatusSuccesful();
$this->assertJsonContains($cilentJson);
```

In your ClientJsonMother you will have as many factory methods as it makes sense. If you don't want to generate a random client, but a named one, you might have:

$clientJson = ClientJsonMother:named($name);

For postScripts:
$clientJson = ClientJsonMother:withScripts($preScriptsArray, $postScriptsArray);

But to be honest, I suspect that probably making a "random()" one that sets a timestamped name and valid random values for other properties will probably be enough for most of your tests.

**Comment by elacunza** [comment]
Sure, do it as it fits you. Thanks!

**Pull request closed without merging**
